Add the REST API's prepare_item_for_response() method

wvega opened the pull request:
# Summary

This PR adds the `prepare_item_for_response()` method to the `Settings` REST Controller class.

### Story: [CH 32761](https://app.clubhouse.io/skyverge/story/32761/add-the-rest-api-prepare-item-for-response-method)
### Release: #436

## Details

I wasn't sure whether the implementation document expected the method to return an array, or a REST Response object containing an array. However, since `WP_REST_Controller::prepare_item_for_response()` returns `WP_Error|WP_REST_Response`, I decided to match the parent's method return type in this PR.

From what I can see, we would be converting the array into a response object anyway, so I think it makes sense to match the return type of the parent method.

## QA

- [x] Integration tests pass

wvega commented on tests/integration/REST_API/Controllers/SettingsTest.php:
```diff
@@ -0,0 +1,142 @@
+
+		$item = $controller->prepare_item_for_response( $settings->get_setting( 'test' ), null )->get_data();
+
+		$this->assertEquals( null, $item['value'] );
```
I'm thinking that we should return the default value instead, as in `Abstract_Settings::get_value()`. What do you think?

deivamagalhaes replied:
I am not sure. If we do, we would lose visibility from the API if the value is set or not. Is there any circumstance when we would need to know if the database option is not set or is set to the default value? @ChaseWiseman , thoughts?

ChaseWiseman replied:
I think in the REST context, it'd be valuable to differentiate between a setting that's been explicitly set vs. a setting that may have a default but not yet saved. The consumer (e.g. JS app) can decide whether they want to pre-fill their field with the default value or display/use it in some other way.

deivamagalhaes commented on tests/integration/REST_API/Controllers/SettingsTest.php:
```diff
@@ -0,0 +1,142 @@
+
+		$settings->register_control( 'test', Control::TYPE_SELECT, [
+			'name'  => 'Select field',
+			'description' => 'A regultar select input field',
```
```suggestion
			'description' => 'A regular select input field',
```

deivamagalhaes commented on woocommerce/rest-api/Controllers/Settings.php:
```diff
@@ -155,7 +156,44 @@ public function update_item_permissions_check( $request ) {
 
-	// TODO: prepare_item_for_response()
+	/**
+	 * @sincex.y.z
```
```suggestion
	 * @since x.y.z
```

deivamagalhaes commented on woocommerce/rest-api/Controllers/Settings.php:
```diff
@@ -155,7 +156,44 @@ public function update_item_permissions_check( $request ) {
+			];
+
+			if ( $control = $setting->get_control() ) {
+				$item['control'] = [
```
I think it may be better for consistency to have the control key with a null value then not to have it.

wvega replied:
Sounds good. I changed that in 6249c31 and 5499ec1

deivamagalhaes commented on woocommerce/rest-api/Controllers/Settings.php:
```diff
@@ -155,7 +156,44 @@ public function update_item_permissions_check( $request ) {
+	 *
+	 * @param Setting $setting a setting object
+	 * @param \WP_REST_Request $request request object
+	 * @return WP_Error|WP_HTTP_Response
```
Shouldn't this be `WP_REST_Response`?

wvega replied:
Yes. Good catch!

deivamagalhaes commented on woocommerce/rest-api/Controllers/Settings.php:
```diff
@@ -155,7 +156,44 @@ public function update_item_permissions_check( $request ) {
+	 *
+	 * @param Setting $setting a setting object
+	 * @param \WP_REST_Request $request request object
+	 * @return WP_Error|WP_HTTP_Response
```
```suggestion
	 * @return WP_Error|WP_REST_Response
```

deivamagalhaes commented on tests/integration/REST_API/Controllers/SettingsTest.php:
```diff
@@ -0,0 +1,142 @@
+	/** Tests *********************************************************************************************************/
+
+
+	/** @see Abstract_Settings::get_id() */
```
Is this correct?

wvega replied:
Noup. I left lots of copypasta on this class :(. Fixed in a717049

deivamagalhaes commented on tests/integration/REST_API/Controllers/SettingsTest.php:
```diff
@@ -0,0 +1,142 @@
+	}
+
+
+	/** @see Abstract_Settings::get_id() */
```
Same here.

deivamagalhaes commented on tests/integration/REST_API/Controllers/SettingsTest.php:
```diff
@@ -0,0 +1,142 @@
+	}
+
+
+	/** @see Abstract_Settings::get_id() */
```
Same here

deivamagalhaes commented on tests/integration/REST_API/Controllers/SettingsTest.php:
```diff
@@ -0,0 +1,142 @@
+use SkyVerge\WooCommerce\PluginFramework\v5_6_1\Settings_API\Control;
+
+/**
+ * Tests for the Abstract_Settings class.
```
This one also needs to be updated.

wvega commented:
@deivamagalhaes thank you for reviewing. I'm sorry for all the copypasta and typos 🙏

I think I already addressed all your feedback. Would you mind taking another look?

deivamagalhaes commented on tests/integration/REST_API/Controllers/SettingsTest.php:
```diff
@@ -0,0 +1,142 @@
+
+		$item = $controller->prepare_item_for_response( $settings->get_setting( 'test' ), null )->get_data();
+
+		$this->assertEquals( null, $item['value'] );
```
Maybe a comment above this line to document our decision?

wvega replied:
Good call. I added one in fa6bfdb.
